cdap-master/src/main/java/co/cask/cdap/data/startup/FileSystemCheck.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data.startup;
+
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import com.google.inject.Inject;
+import org.apache.twill.filesystem.Location;
+import org.apache.twill.filesystem.LocationFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Checks that the configured FileSystem is available and the root path has the correct permissions.
+ */
+// class is picked up through classpath examination
+@SuppressWarnings("unused")
+public class FileSystemCheck extends BaseMasterCheck {
+  private static final Logger LOG = LoggerFactory.getLogger(FileSystemCheck.class);
+  private final LocationFactory locationFactory;
+
+  @Inject
+  private FileSystemCheck(CConfiguration cConf, LocationFactory locationFactory) {
+    super(cConf);
+    this.locationFactory = locationFactory;
+  }
+
+  @Override
+  public void run() throws Exception {
+    String hdfsUser = cConf.get(Constants.CFG_HDFS_USER);
+    String hdfsRootPath = cConf.get(Constants.CFG_HDFS_NAMESPACE);
+    LOG.info("Checking that user {} has permission to write to {} on the FileSystem.",
+             hdfsUser, hdfsRootPath);
+
+    Location rootLocation = locationFactory.create(hdfsRootPath);
+    try {
+      if (!rootLocation.mkdirs()) {
```

Comment:
hm took a closer look and LocalLocation will return false, but HDFSLocation will return true, due to differences in how FileSystem and File implement mkdirs().  I will check for existence first then.